File: solutions/ruby/crypto-square/1/crypto_square.rb
class Crypto
  def initialize(plaintext)
    @plaintext = plaintext
  end

  def ciphertext
    normalized = @plaintext.downcase.gsub(/[^a-z0-9]/, '')
    return '' if normalized.empty?

    l = normalized.length
    c = Math.sqrt(l).ceil
    r = (l.to_f / c).ceil

    rows = normalized.chars.each_slice(c).map(&:join)
    rows[-1] = rows[-1].ljust(c) if rows.any?
    
    # 実際には r 行必要だが、最後の行が足りない場合がある
    # columns は c 個必要。各 column は r 文字。
    
    padded = normalized.ljust(r * c)
    rect = padded.chars.each_slice(c).to_a
    
    rect.transpose.map(&:join).join(' ')
  end
end

File: solutions/javascript/crypto-square/1/crypto-square.js
export class Crypto {
  constructor(text) {
    this._normalized = text.toLowerCase().replace(/[^a-z0-9]/g, '');
  }

  get ciphertext() {
    if (this._normalized === '') return '';
    const len = this._normalized.length;
    const cols = Math.ceil(Math.sqrt(len));
    const rows = Math.ceil(len / cols);
    const result = [];
    for (let c = 0; c < cols; c++) {
      let chunk = '';
      for (let r = 0; r < rows; r++) {
        const idx = r * cols + c;
        chunk += idx < len ? this._normalized[idx] : ' ';
      }
      result.push(chunk);
    }
    return result.join(' ');
  }
}
